gquant/dataframe_flow/_node_flow.py

```diff
@@ -32,6 +29,40 @@
 #     to_port = 'to_port'
 #     from_port = 'from_port'
 
+_validators = {}  # dictionary of validators of funciton signiture (val, meta) -> bool
+
+_copys = {}  # dictionary of object copy functions
+
+_cleanup = {}  # dictionary of clean up functions
+
+
+def dynamic_ports(ports_setup):
+
+    def calculated_ports_setup(self):
```

Comment:
I feel like this `dynamic_ports` and `calculated_ports_setup` is unnecessary to be present in the `Node` class. Instead just add an overriding `ports_setup` method in the `NodeTaskGraphMixin` class. The following code I think will work:

```
in class NodeTaskGraphMixin

def ports_setup(self):
    nodecls_list = _get_nodetype(self)  # this will filter out the primary class with ports_setup
    for icls in nodecls_list:
        if not hasattr(icls, 'ports_setup'):
            continue
        ports = icls.ports_setup(self)
        break
    else:
        raise Exception('ports_setup method missing')

    port_type = PortsSpecSchema.port_type
    inports = ports.inports
    dy = PortsSpecSchema.dynamic
    for key in inports:
        if dy in inports[key] and inports[key][dy]:
            types = inports[key][port_type]
            break
    else:
        return ports
    if hasattr(self, 'inputs'):
        for inp in self.inputs:
            to_port = inp['to_port']
            if to_port in inports and (not inports[to_port].get(dy,
                                                                False)):
                # skip connected non dynamic ports
                continue
            else:
                inports[inp['from_node'].uid+'@'+inp['from_port']] = {
                    port_type: types, dy: True}
    return ports


```

This way you can just keep using `ports_setup` everywhere.

Reply:
yeah. if this works, it will be better. Let me try it.

Reply:
Yes. it works well